metricbeat/module/system/diskio/diskstat_windows_helper.go

```diff
@@ -0,0 +1,169 @@
+	WriteCount          uint32
+	QueueDepth          uint32
+	SplitCount          uint32
+	QueryTime           int64
+	StorageDeviceNumber uint32
+	StorageManagerName  [8]uint16
+}
+
+// used to get the specific IOCTL_DISK_PERFORMANCE control code
+func getCtlCode(deviceType uint32, function uint32, method uint32, access uint32) uint32 {
+	return (deviceType << 16) | (access << 14) | (function << 2) | method
+}
+
+// IOCounters gets the diskio counters and maps them to the list of counterstat objects
+func iOCounters(names ...string) (map[string]disk.IOCountersStat, error) {
+	ret := make(map[string]disk.IOCountersStat, 0)
+	logicalDisks, err := getLogicalDriveStrings()
+	if err != nil || len(logicalDisks) == 0 {
+		return nil, err
+	}
+	for _, drive := range logicalDisks {
+		if len(drive.Name) > 3 { // not get _Total or Harddrive
+			continue
+		}
+
+		if len(names) > 0 && !containsDrive(names, drive.Name) { //filter by included devices
+			continue
+		}
+
+		var counter, err = iOCounter(drive.UNCPath)
+		if err != nil {
+			return nil, err
+		}
+		ret[drive.Name] = disk.IOCountersStat{
+			Name:       drive.Name,
+			ReadCount:  uint64(counter.ReadCount),
+			WriteCount: uint64(counter.WriteCount),
+			ReadBytes:  uint64(counter.BytesRead),
+			WriteBytes: uint64(counter.BytesWritten),
+			ReadTime:   uint64(counter.ReadTime),
+			WriteTime:  uint64(counter.WriteTime),
+		}
+	}
+	return ret, nil
+}
+
+// iOCounter calls syscal func CreateFile to generate a handler then executes the DeviceIoControl func in order to retrieve the metrics
+func iOCounter(path string) (diskPerformance, error) {
+	var diskPerformance diskPerformance
+	var diskPerformanceSize uint32
+	utfPath, err := syscall.UTF16PtrFromString(path)
+	if err != nil {
+		return diskPerformance, err
+	}
+	hFile, err := syscall.CreateFile(utfPath,
+		syscall.GENERIC_READ|syscall.GENERIC_WRITE,
+		syscall.FILE_SHARE_READ|syscall.FILE_SHARE_WRITE,
+		nil,
+		syscall.OPEN_EXISTING,
+		syscall.FILE_FLAG_BACKUP_SEMANTICS,
+		0)
+
+	if err != nil {
+		return diskPerformance, err
+	}
+	defer syscall.CloseHandle(hFile)
+
+	err = syscall.DeviceIoControl(hFile,
+		ioctlDiskPerformance,
+		nil,
+		0,
+		(*byte)(unsafe.Pointer(&diskPerformance)),
+		uint32(unsafe.Sizeof(diskPerformance)),
+		&diskPerformanceSize,
+		nil)
+	if err != nil {
+		return diskPerformance, err
+	}
+	return diskPerformance, nil
+}
+
+// getLogicalDriveStrings calls the syscall GetLogicalDriveStrings in order to get the list of logical drives
+func getLogicalDriveStrings() ([]logicalDrive, error) {
+	lpBuffer := make([]byte, 254)
+	logicalDrives := make([]logicalDrive, 0)
+	r1, _, e1 := syscall.Syscall(procGetLogicalDriveStringsW.Addr(), 2, uintptr(len(lpBuffer)), uintptr(unsafe.Pointer(&lpBuffer[0])), 0)
+	if r1 == 0 {
+		if e1 != errorSuccess {
+			return nil, e1
+		} else {
+			return nil, syscall.EINVAL
+		}
+	}
+
+	for _, v := range lpBuffer {
+		if v >= 65 && v <= 90 {
+			path := string(v) + ":"
```

Comment:
What about makint that modification on line 154 instead, so the comparison below gets simpler and we don't need to create a string in case of continue?